cmd/notation/registry.go

```diff
@@ -55,6 +59,56 @@ func getRepositoryClient(ctx context.Context, opts *SecureFlagOpts, ref registry
 	return notationregistry.NewRepository(remoteRepo), nil
 }
 
+// getSignatureRepositoryForSign returns a registry.Repository for Sign.
+// ociImageManifest denotes the type of manifest used to store signatures during
+// Sign process.
+// Setting ociImageManifest to true means using OCI image manifest.
+// Otherwise, use OCI artifact manifest and requires Referrers API.
+func getSignatureRepositoryForSign(ctx context.Context, opts *SecureFlagOpts, reference string, ociImageManifest bool) (notationregistry.Repository, error) {
+	ref, err := registry.ParseReference(reference)
+	if err != nil {
+		return nil, err
+	}
+
+	// generate notation repository
+	return getRepositoryClientForSign(ctx, opts, ref, ociImageManifest)
+}
+
+func getRepositoryClientForSign(ctx context.Context, opts *SecureFlagOpts, ref registry.Reference, ociImageManifest bool) (notationregistry.Repository, error) {
+	authClient, plainHTTP, err := getAuthClient(ctx, opts, ref)
+	if err != nil {
+		return nil, err
+	}
+
+	remoteRepo := &remote.Repository{
+		Client:    authClient,
+		Reference: ref,
+		PlainHTTP: plainHTTP,
+	}
+	repositoryOpts := notationregistry.RepositoryOptions{
+		OCIImageManifest: ociImageManifest,
+	}
+	// using OCI artifact manifest to store signatures. This requires the
+	// existence of Referrers API for Sign process.
+	if !ociImageManifest {
```

Comment:
Could you help me to understand the scenarios here? Should we list the valid scenarios in the comment? 
From the code it looks like:
1. If OCI artifact manifest is used
  - Referrers API must be available
  - Must not fallback to referrers tag schema (why?)

2. If OCI Image manifest is used
  - Referrers API and tag schema are both OK (?)

Reply:
This change is based on this community [discussion](https://cloud-native.slack.com/archives/CQUH8U287/p1674050115958189?thread_ts=1673362612.471549&cid=CQUH8U287) that notation will not implement hybrid support, for instance, a registry supports OCI artifact manifest but not the Referrers API. 
Therefore, when user wants to store signatures in OCI artifact manifest, we require the Referrers API as well (and this is the default behavior of notation). 
On the other hand, I haven't seen a registry that supports Referrers API but only OCI image manifest. Hence, when user chooses OCI image manifest, notation could utilize oras's fallback logic and use the tag schema. But to be safer, yes, I think setting `remoteRepo.SetReferrersCapability(false)` on OCI image manifest is a better way here. I will update the code to reflect this.

Reply:
Thanks @patrickzheng200  for the explanation.

> But to be safer, yes, I think setting remoteRepo.SetReferrersCapability(false) on OCI image manifest is a better way here.

To my understanding, it seems that it won't hurt to use the Referrers API for OCI image manifests. But if the Referrers API is known to be unavailable, setting `remoteRepo.SetReferrersCapability(false)` will bring better performance since the call to Referrers API will be saved.

Reply:
> Thanks @patrickzheng200 for the explanation.
> 
> > But to be safer, yes, I think setting remoteRepo.SetReferrersCapability(false) on OCI image manifest is a better way here.
> 
> To my understanding, it seems that it won't hurt to use the Referrers API for OCI image manifests. But if the Referrers API is known to be unavailable, setting `remoteRepo.SetReferrersCapability(false)` will bring better performance since the call to Referrers API will be saved.

Just had another discussion with Shiwei, Notation would still enforce two paths during Signing:
1. OCI artifact manifest uses the Referrers API
2. OCI image manifest uses the Referrers Tag Schema
(So basically, we do not support hybrid case on both ways.)

Reply:
Got it!